### ISSUE TITLE
Write new base Application class

### ISSUE THREAD

**Issue by m-schubert** [alt-lu/pywellcad#113]
The entrypoint of an `Application` class makes sense for the API, but the interface could be a lot more Pythonic. We should spike out a sensible `Application` class as the base of the non-COM API. Some suggestions:

- Give the ability to connect to an existing running WellCAD process
- Allow opening of new WellCAD process
- Access to `BoreholeDocument` via a smart container that allows iteration, searching by name, etc:
  ```python
  app = Application()
  well_x = app.borehole_documents.named("Well X")
  print(app.borehole_documents.active.name)
  for bd in app.borehole_documents:
      print(bh.name)
  ```
- Sensible file importing from TFD, LAS, DLIS, TXT, CSV, etc. In the COM interface it is one single method, but perhaps it should be split. Possibly it should go on the `BoreholeDocument` class itself and we then attach a `BoreholeDocument` to the `Application`?
  ```python
  app = Application()
  app.insert(BoreholeDocument.open_csv('C:/Temp/foo.csv'))
  ```
- Standard application control like minimize, maximize, close, etc.

**Comment by m-schubert** [comment]
It's worth taking a look at #110 as well. I suspect `BoreholeDocumentCollection` might be the first smart container to implement this.

**Comment by janwillembuist** [comment]
Hi!
I'm currently trying out `pywellcad` and really like it so far!

Adding on to this, maybe it would be an idea to make the application class itself by default an iterator over boreholes and the borehole class an iterator over the logs.

```python
app = Application()

for borehole in app:
    for log in borehole:
        print(log)

```

**Comment by m-schubert** [comment]
> Hi! I'm currently trying out `pywellcad` and really like it so far!
> 
> Adding on to this, maybe it would be an idea to make the application class itself by default an iterator over boreholes and the borehole class an iterator over the logs. Typically, `pywellcad` would be something I would use to do a lot of normally repetitive tasks, so adding iterator behavior would be quite helpful.
> 
> For example:
> 
> ```python
> app = Application()
> 
> for borehole in app:
>     for log in borehole:
>         print(log)
> ```
> 
> Let me know what you think, I am happy to open a PR!

Great idea. Check out #110 - it's a more specific issue for this iterator case

Please go ahead and open a PR. Just one thing to note - we'd like to keep the `wellcad.com` module as a vanilla implementation of the COM interface exposed by the Automation API. If you'd like to extend `Application` and `Borehole`, go ahead and do it directly in the `wellcad` module.

**Comment by janwillembuist** [comment]
Hi I was just looking into it a little bit. I understand you want to keep things separate, however this makes the job quite tedious as you have to override all the methods inside `Application` that initialize a `Borehole`, with extended version of `Borehole`. Same goes for `Log`. If you then later decide to make changes you have to keep track of everything.

Instead, I would say add a simple `__iter__` method to `Application`, `Borehole`, and `Log`. For example, in `Application`:
```python
    def __iter__(self):
        return iter([self.get_borehole(i) for i in range(self.nb_of_documents)])
```

**Comment by m-schubert** [comment]
Yeah, that's a good point. OK, get a PR for `__iter__` and we'll merge it. Just by the way, you can reduce your example to just a generator expression and avoid constructing a list:

```python
    def __iter__(self):
        return (self.get_borehole(i) for i in range(self.nb_of_documents))
```
